bitcoin/src/lib.rs

```diff
@@ -42,12 +42,19 @@
 #![allow(clippy::manual_range_contains)] // More readable than clippy's format.
 #![allow(clippy::needless_borrows_for_generic_args)] // https://github.com/rust-lang/rust-clippy/issues/12454
 
-// Disable 16-bit support at least for now as we can't guarantee it yet.
-#[cfg(target_pointer_width = "16")]
-compile_error!(
-    "rust-bitcoin currently only supports architectures with pointers wider than 16 bits, let us
-    know if you want 16-bit support. Note that we do NOT guarantee that we will implement it!"
-);
+use core::mem;
+
+// We only support machines with index size of 4 or 8 bytes.
+//
+// Bitcoin consensus code relies on being able to have containers with more than 65536 (2^16)
+// entries in them so we cannot support consensus logic on machines that only have 16-bit memory
+// addresses.
+//
+// We specifically do not use `target_pointer_width` because of the possibility that pointer width
+// does not equal index size.
+//
+// ref: https://github.com/rust-bitcoin/rust-bitcoin/pull/2929#discussion_r1661848565
+internals::const_assert!(mem::size_of::<usize>() == 4 || mem::size_of::<usize>() == 8);
```

Comment:
This again is far from the `ToU64` trait which makes it hard to review and keep track of. The `ToU64 for usize` impl should be feature gated (so that `units` keep working) and the condition should be put there. Conversely we should have `ToUsize` trait doing the opposite. Finally the conditions should be `mem::size_of::<usize> <= 8` for `ToU64` and `mem::size_of::<usize> >= 4` for `ToUsize`. I'm also tempted to do two feature gates in case some crate only needs to require one condition but I suspect that won't happen in practice since encoding mirrors decoding.

Reply:
> This again is far from the ToU64 trait which makes it hard to review and keep track of.

I don't quite understand this, we discussed that `ToU64` is not the only reason we have this restriction, the other reason (the consensus stuff) is general to the library so therefore the check is crate wide and at the top of `lib.rs` - what am I misunderstanding? 

Can use the conditions as suggested.

> I'm also tempted to do two feature gates

This might actually be better because the ">= 4" is because of consensus and the "<= 8" is because of `ToU64`, right?

Reply:
The reason we need `ToU64` in the first place is consensus stuff. We should rewrite all casts to use it. (Eventually, not really all at once.)

> the ">= 4" is because of consensus and the "<= 8" is because of `ToU64`, right?

Consensus rules require both. Varint can't encode sizes above 2<sup>64</sup> and there are transactions in Bitcoin that have more than 2<sup>16</sup> items. The reason I'm tempted to split them is that it's cleaner and we may find that there is a place with just one of them somewhere.

Reply:
If we want to enforce that nothing has size exceeding 2^64 (which seems a little silly and unnecessary because of the impracticality of creating such objects, but ok) then we ought to do it by making sure that all of our data structures have APIs that enforce this invariant.

Personally I think we should just let it go. We can panic if we try to varint-encode a number that's too high. Or saturate, or encode nothing, or whatever. It won't ever happen.

Making the code literally impossible to compile on CHERI or something seems like the wrong approach.

Reply:
> Making the code literally impossible to compile on CHERI or something seems like the wrong approach.

That's not really what's happening - it has 128-bit pointers, not sizes. We'll bounding by sizes.

Reply:
As an anecdote, my work uses `u64::try_from().unwrap()` as yes, that'd require an impossible amount of RAM to fail _when a buffer length_ and a system whose usize exceeds u64. I don't mind enforcing usize <= 64-bit, and that should not break CHERI, but I think it's sufficiently edge case it doesn't need action _in my own work_..

Reply:
Ok, I rescind my complaint and accept having the "usize must be 4 or 8 bytes" condition.